### ISSUE TITLE
slow calls to elastic, and wondering about need for the -- clock.sleep(TimeUnit.SECONDS.toMillis(1)) 

### ISSUE THREAD

**Issue by nbuesing** [confluentinc/kafka-connect-elasticsearch#487]
I am having some serious performance issues with the latest elastic sink but it is on some new code I am working on, so I am not sure if it is related to the latest version or not, but based on the code change I did to verify, I believe it is.

In the ElasticsearchClient there are two spots where clock.sleep of one second is performed.  Looks like while there are attempts to increase buffer size, there is a hardcoding 1 second sleep.  

Locally I changed the TimeUnit from Seconds to Milliseconds and the code is being performant.  Oddly, I do not always see this behavior, but I think it could be related to the amount of updates vs inserts (in addition to batch sizes). 
  

```
    if (docIdToRecord != null) {
      // make sure that only unique records are being sent in a batch
      // every request that is flushed and succeeds triggers a callback that removes it from the map
      while (docIdToRecord.containsKey(request.id())) {
        flush();
        clock.sleep(TimeUnit.MILLISECONDS.toMillis(1));
      }
    }

    // wait for internal buffer to be less than max.buffered.records configuration
    long maxWaitTime = clock.milliseconds() + config.flushTimeoutMs();
    while (numRecords.get() >= config.maxBufferedRecords()) {
      clock.sleep(TimeUnit.MILLISECONDS.toMillis(1));
      if (clock.milliseconds() > maxWaitTime) {
        throw new ConnectException(
            String.format(
                "Could not make space in the internal buffer fast enough. Consider increasing %s"
                    + " or %s.",
                FLUSH_TIMEOUT_MS_CONFIG,
                MAX_BUFFERED_RECORDS_CONFIG
            )
        );
      }
    }
```
CC-12445: Replace the JestClient with the ElasticsearchRestClient
https://github.com/confluentinc/kafka-connect-elasticsearch/pull/468

**Comment by levzem** [comment]
@nbuesing I agree, this seems to be a concern for data sets with frequent updates to documents

I have published a PR #490 that address this issue and will follow up with a bug fix release shortly

**Comment by nbuesing** [comment]
in comparing to 10.0.2 which I am told would be compatible with connect cluster 6.0.x, there is no issue.  Also it looks as though the flushing to avoid duplicates in a bulk isn't done "1 at a time", 

In 11.0.x version we get tons of lines like this: WARN Ignoring version conflict for operation INDEX on document '8f1a0cab-5b8c-11eb-a030-f53c337f4933' one at a time.

In 10.0.2 we get longer messages (and less of them) 

WARN Ignoring version conflicts for items: [Key{entity-X/_doc/51a8c5c9-5b92-11eb-a030-1b6c413695e2}, Key{entity-X/_doc/111c7912-55f1-11eb-9b5b-cb91b21ff31d}, Key{entity-X/_doc/d1659053-55f3-11eb-9b5b-8b23b862f6a1}, ....

It looks like duplicates were handled in bulk not individually 

I worry that in extreme cases reducing the sleep from 1s to 10ms is still not performant; especially when the previous connector shows no signs of having this behavior.

In my custom version of 11.0.0 where I changed the sleep from 1s to 1ms, the connector is dramatically slower than version 10.0.2

**Comment by levzem** [comment]
@nbuesing 11.0.x merely unwraps each individual response `https://github.com/confluentinc/kafka-connect-elasticsearch/blob/master/src/main/java/io/confluent/connect/elasticsearch/ElasticsearchClient.java#L321-L324`

while 10.0.x also unwraps each individual record and chooses to report all of the conflicts in one line
https://github.com/confluentinc/kafka-connect-elasticsearch/blob/10.0.x/src/main/java/io/confluent/connect/elasticsearch/jest/JestElasticsearchClient.java#L607-L623

hence the longer and less frequent messages

there is no difference how the version conflicts are handled besides how they are logged.

the reason I changed it to 10ms was because I wanted to give back some time to the CPU as requests do not complete in 1ms. i need to test the difference but I would not be surprised if there is next to none between 1ms and 10ms

as for poorer performance, I will have to look into that some more. I would bet it is only poorer in performance for updates and not general inserts because no behavior has changed with respect to that, unless the RestHighLevelClient bulk processor is significantly slower than the one we replaced

**Comment by nbuesing** [comment]
Thanks for the information, here is my observations that I completed this evening.  I will have to look into RestHighLevelClient to see why that would be slower -- at least in my test here I think I have to keep all the invariants constant between the two tests.

Thanks

--
a performance test comparing 10.0.2 against 11.0.1 (1s sleep changed to 1ms)

total number of events in the search-x = 3,884,301

stats:
connector version=10.0.2
rate=2800/s
duration=22:30 minutes
unique records-in-elastic=387035

status:
connector version=11.0.1 (11.0.0 version with 1s sleep replaced with 1ms)
rate=800/s
duration= 1:20:00 hours
unique records-in-elastic=387035

which puts it 3.5x slower -- same topic data (just reset consumer group to earliest) and same connector settings.

settings:
1 topic / 1 elastic index / 1 connect cluster / 1 connect worker
poll=500ms, batch.size=100, max.in.flight.requests=5,fetch.max.wait.ms=5000
key.ignore=false, schema.ignore=false, behavior.on.null.values=delete

I didn't notice the 11.0.0 connector died, so time provided excludes the 10 minutes that went unnoticed:
-10 minutes for a connection 
Caused by: org.apache.kafka.connect.errors.ConnectException: Bulk request failed.
        at io.confluent.connect.elasticsearch.ElasticsearchClient$1.afterBulk(ElasticsearchClient.java:337)

**Comment by levzem** [comment]
@nbuesing I have update the PR to remove the wait and the unique key constraint

I anticipate this should solve the performance issue. I will run some performance tests of my own to confirm

**Comment by levzem** [comment]
@nbuesing `v11.0.1` is now out which should address the issue

**Comment by nbuesing** [comment]
@levzem, thanks

**Comment by levzem** [comment]
closing as this was fixed in #490
